### PR TITLE
Tile & Splashscreen generator

### PR DESCRIPTION
**Pull request by davesmits**
The long promised tile / splashscreen generator #616

### CODE REVIEW

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+
+        private Vector2 GetWSAImageTypeSize(PlayerSettings.WSAImageType type, PlayerSettings.WSAImageScale scale)
+        {
+            Vector2 size = Vector2.zero;
+            float scaleFactor = (float)scale / 100;
+
+            if (scale == PlayerSettings.WSAImageScale._125)
+            {
+
+            }
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: what's this empty if statement for?

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+using System.Reflection;
+using UnityEditor;
+using UnityEngine;
+
+namespace HoloToolkit.Unity
+{
+
+    public class SetIconsWindow : EditorWindow
+    {
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: nit: extra lines after brackets for namespace & class

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+            this.titleContent = new GUIContent(WindowTitle);
+            this.minSize = new Vector2(600, 200);
+
+            // Load settings
+            _originalAppIconPath = EditorPrefs.GetString(EditorPrefsKey_AppIcon);
+            _originalSplashImagePath = EditorPrefs.GetString(EditorPrefsKey_SplashImage);
+            _outputDirectoryName = EditorPrefs.GetString(EditorPrefsKey_DirectoryName);
+
+            if (!string.IsNullOrEmpty(_originalAppIconPath))
+                _originalAppIcon = AssetDatabase.LoadAssetAtPath<Texture2D>(_originalAppIconPath);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: nit: please use brackets in all if statements

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+
+        private string CloneAndResizeToFile(Texture2D texture, Vector2 iconSize, ref string fileName)
+        {
+            string directoryPath = Application.dataPath + "/" + _outputDirectoryName;
+            string filepath = directoryPath + "/" + fileName;
+
+            if (!((iconSize.x == 1240 && iconSize.y == 1240) ||
+                (iconSize.x == 2480 && iconSize.y == 1200)))
+            {
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: nit: extra line space

File: Assets/HoloToolkit/Utilities/Editor/TextureScale.cs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            finishCount++;
+            mutex.ReleaseMutex();
+        }
+
+        private static Color ColorLerpUnclamped(Color c1, Color c2, float value)
+        {
+            return new Color(c1.r + (c2.r - c1.r) * value,
+                              c1.g + (c2.g - c1.g) * value,
+                              c1.b + (c2.b - c1.b) * value,
+                              c1.a + (c2.a - c1.a) * value);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/TextureScale.cs]: nit: formatting is off by one space to the right

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+                case PlayerSettings.WSAImageType.StoreTileWideLogo:
+                case PlayerSettings.WSAImageType.UWPWide310x150Logo:
+                    if (scale != PlayerSettings.WSAImageScale.Target16 &&
+                        scale != PlayerSettings.WSAImageScale.Target24 &&
+                        scale != PlayerSettings.WSAImageScale.Target32 &&
+                        scale != PlayerSettings.WSAImageScale.Target48 &&
+                        scale != PlayerSettings.WSAImageScale.Target256)
+                        return _originalSplashImage;
+                    else
+                        break;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: nit: brackets around this if else, please

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+                        if (texture != null)
+                        {
+                            string filename = null; //string.Format("{0}{1}.png", type, scale);
+                            CloneAndResizeToFile(texture, type, scale, ref filename);
+                            PlayerSettings.WSA.SetVisualAssetsImage(string.Format("Assets/{0}/{1}", _outputDirectoryName, filename), type, scale);
+
+                            progress++;
+                            cancelled = EditorUtility.DisplayCancelableProgressBar("Generating images", string.Format("Generating resized images {0} of {1}", progress, progressTotal), progress / progressTotal);
+                            if (cancelled)
+                                break;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: nit: brackets

**Comment by HoloFan** [comment]
Hey
I tried using this extensions as it sounds wonderful to easily create icons and splashscreens.

In order to use it I copied the files into my asset foulder and in the top toolbar the generator showed up as excpected.
I opened it, added the image in the correct size and hit the "Update Icon"-Button but nothing happened. Am I doing something wrong or missing out?

Thanks in advance!

**Comment by GKersten** [comment]
@HoloFan Love yo hear that you are already trying out our extension!:) As this is still quite a new extension.. we should add a check if the output folder already exists, so at this point there is no check.. What you should do is add a folder to your Assets folder with a given name "AppIcons" for example. Then copy this folder name to the "Output folder" field in the extension. That should work! You will see a progressbar indicating the generation of the icons!

**Comment by HoloFan** [comment]
@GKersten: I was able to succesfully generate the Icon and Splashscreen images. I just noticed, that the Splashscreen images does not get set for Windows Holographic in the Player settings. I set it manually and used the 2480x1200 version which worked just fine.

**Comment by GKersten** [comment]
@HoloFan I will take a look at this, but I've also noticed that Unity does not always update the Player Settings correctly in the UI, but the settings are actually set correctly. If you close the player settings and reopen the windows, or sometimes even restart Unity, you will see the image is actually set.

**Comment by HoloFan** [comment]
@GKersten : Ok I actually build the app and tried it with my HoloLens and no Splashscrren was shown. That's how I noticed that it was not set. But maybe I would have restarted Unity once and then build the app it would have worked. Who knows :D

**Comment by Ziugy** [comment]
@HoloFan: Try restarting your HoloLens device. A co-worker saw a similar issue and that was a work-around for this issue. Every time the icon or splash screen was updated this needed to be done. Hopefully that helps.

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+
+            if (!((iconSize.x == 1240 && iconSize.y == 1240) || (iconSize.x == 2480 && iconSize.y == 1200)))
+            {
+
+                // Create copy of original image
+                var clone = CloneAndResizeToTexture(texture, iconSize);
+
+                // Write clone to assets folder
+                var bytes = clone.EncodeToPNG();
+                Directory.CreateDirectory(directoryPath);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: nit: `if(!Directory.Exists(directoryPath))`

**Comment by davesmits** [comment]
fixed the bug and solved your comment

**Comment by davesmits** [comment]
how is going be the second reviewer?

**Comment by StephenHodgson** [comment]
@Ziugy, or @Holofan can both report back on the updates you've made.

File: Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+        public static void OpenWindow()
+        {
+            SetIconsWindow window = GetWindow<SetIconsWindow>(WindowTitle) as SetIconsWindow;
+            if (window != null)
+            {
+                window.Show();
+            }
+        }
+
+        public void OnEnable()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/SetIconsWindow.cs]: Nit: Unity methods should be private.  Goes for OnDisable and OnGUI too

File: Assets/HoloToolkit/Utilities/Editor/TextureScale.cs [hunk 10]
```diff
@@ -0,0 +1,160 @@
+﻿
+using System.Threading;
+using UnityEngine;
+
+namespace HoloToolkit.Unity
+{
+    public class TextureScale
+    {
+        public class ThreadData
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/Utilities/Editor/TextureScale.cs]: Nit, should this class be its own file?

**Comment by davesmits** [comment]
thanks! just processed your feedback

**Pull request merged**